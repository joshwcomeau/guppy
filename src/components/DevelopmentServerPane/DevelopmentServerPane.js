// @flow
import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { launchDevServer, abortTask } from '../../actions';
import { getSelectedProject } from '../../reducers/projects.reducer';
import { getDevServerTaskForProjectId } from '../../reducers/tasks.reducer';
import { getDocumentationLink } from '../../services/project-type-specifics';
import { BREAKPOINTS, GUPPY_REPO_URL } from '../../constants';

import Module from '../Module';
import Card from '../Card';
import Toggle from '../Toggle';
import Spacer from '../Spacer';
import TerminalOutput from '../TerminalOutput';
import ExternalLink from '../ExternalLink';
import OnlyOn from '../OnlyOn';
import DevelopmentServerStatus from '../DevelopmentServerStatus';

import type { Project, Task } from '../../types';

type Props = {
  project: Project,
  task: ?Task,
  launchDevServer: (task: Task, timestamp: Date) => void,
  abortTask: (task: Task, timestamp: Date) => void,
};

class DevelopmentServerPane extends PureComponent<Props> {
  handleToggle = (isToggled: boolean) => {
    const { task, launchDevServer, abortTask } = this.props;

    if (!task) {
      // Should be impossible, since the Toggle control won't render without
      // a "start" task.
      return;
    }

    const timestamp = new Date();

    if (isToggled) {
      launchDevServer(task, timestamp);
    } else {
      abortTask(task, timestamp);
    }
  };

  render() {
    const { project, task } = this.props;

    if (!task) {
      // For the initial paint, there won't be any tasks, as tasks aren't
      // persisted and need to be read from the disk.
      // TODO: Differentiate between "the tasks haven't loaded" and "there
      // are no tasks"
      // TODO: Add "skeleton" structure to make it clear it's loading,
      // and to prevent content from jumping around.
      return null;
    }

    // TODO: There's currently no DevelopmentServerStatus for smaller windows.
    // Create a custom component for windows <900px wide

    const isRunning = task.status !== 'idle';

    const description = (
      <Description>
        Runs a local development server that updates whenever you make changes
        to the files.
      </Description>
    );

    const docLink = (
      <DocumentationLink>
        <ExternalLink href={getDocumentationLink(project.type)}>
          View Documentation
        </ExternalLink>
      </DocumentationLink>
    );

    return (
      <Module
        title="Development Server"
        moreInfoHref={`${GUPPY_REPO_URL}/blob/master/docs/getting-started.md#development-server`}
        primaryActionChildren={
          <Toggle isToggled={isRunning} onToggle={this.handleToggle} />
        }
      >
        <OnlyOn size="mdMin">
          <Wrapper>
            <InfoWrapper>
              {description}
              <DevelopmentServerStatus status={task.status} port={task.port} />
              {docLink}
            </InfoWrapper>
            <TerminalWrapper>
              <TerminalOutput height={300} title="Server Logs" task={task} />
            </TerminalWrapper>
          </Wrapper>
        </OnlyOn>

        <OnlyOn size="sm">
          <Wrapper>
            <InfoWrapper>
              <SmallInfoWrapper>
                {description}
                <Spacer size={10} />
                {docLink}
              </SmallInfoWrapper>
            </InfoWrapper>
            <TerminalWrapper>
              <TerminalOutput height={300} task={task} />
            </TerminalWrapper>
          </Wrapper>
        </OnlyOn>
      </Module>
    );
  }
}

const Wrapper = Card.extend`
  display: flex;

  @media ${BREAKPOINTS.sm} {
    flex-direction: column;
  }
`;

const InfoWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;

  @media ${BREAKPOINTS.sm} {
    flex-direction: row;
  }

  @media ${BREAKPOINTS.mdMin} {
    flex: 6;
  }
`;

const SmallInfoWrapper = styled.div`
  display: block;
  padding-bottom: 12px;
`;

const DocumentationLink = styled.div`
  @media ${BREAKPOINTS.mdMin} {
    line-height: 35px;
    text-align: center;
  }
`;

const Description = styled.div`
  font-size: 1.125rem;
`;

const TerminalWrapper = styled.div`
  @media ${BREAKPOINTS.mdMin} {
    flex: 11;
    padding-left: 20px;
    /*
      overflow: hidden is needed so that the column won't expand when the
      terminal output is really long. This way, it will be scrollable.
    */
    overflow: hidden;
    /* Offset by the Card padding amount. */
    margin-top: -15px;
  }
`;

const mapStateToProps = state => {
  const selectedProject = getSelectedProject(state);

  if (!selectedProject) {
    return { task: null };
  }

  return {
    project: selectedProject,
    task: getDevServerTaskForProjectId(
      state,
      selectedProject.id,
      selectedProject.type
    ),
  };
};

const mapDispatchToProps = { launchDevServer, abortTask };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DevelopmentServerPane);
