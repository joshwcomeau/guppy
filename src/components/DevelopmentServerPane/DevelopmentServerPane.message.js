// @flow
import { defineMessages } from 'react-intl';

const messages = {
  main: {
    title: {
      id: 'developmentServerPane.main.title',
      defaultMessage: 'Development Server',
      description: 'Title for the "Development Server" pane',
    },
    noTask: {
      id: 'developmentServerPane.main.noTask',
      defaultMessage:
        'This project does not appear to have a development server task',
      description:
        'Error message when there is no development server task in the project',
    },
    description: {
      id: 'developmentServerPane.main.description',
      defaultMessage:
        'Runs a local development server that updates' +
        ' whenever you make changes to the files.',
      description: 'Description for the development server',
    },
    viewDoc: {
      id: 'developmentServerPane.main.viewDoc',
      defaultMessage: 'View Documentation',
      description: 'Text for "View Documentation"',
    },
  },
};

const definedMessages = defineMessages(messages);

export default definedMessages;
