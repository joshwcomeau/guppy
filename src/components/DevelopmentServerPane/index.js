// @flow
export { default } from './DevelopmentServerPane';
