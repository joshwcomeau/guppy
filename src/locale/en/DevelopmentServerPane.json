{
	"developmentServerPane.main.viewDoc": "View Documentation",
	"developmentServerPane.main.noTask": "This project does not appear to have a development server task",
	"developmentServerPane.main.description": "Runs a local development server that updates whenever you make changes to the files.",
	"developmentServerPane.main.title": "Development Server",
	"developmentServerPane.serverLogs.title" : "Server Logs"
}
