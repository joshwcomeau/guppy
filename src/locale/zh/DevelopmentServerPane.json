{
	"developmentServerPane.main.viewDoc": "查看文档",
	"developmentServerPane.main.noTask": "这个项目似乎没有开发服务器相关的任务",
	"developmentServerPane.main.description": "运行一个在你改变文件时随时更新的本地开发服务器。",
	"developmentServerPane.main.title": "开发服务器",
	"developmentServerPane.serverLogs.title" : "服务器日志"
}
