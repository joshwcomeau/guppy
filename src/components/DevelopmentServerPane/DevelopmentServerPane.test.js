/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import { shallow } from 'enzyme';
import lolex from 'lolex';

import { DevelopmentServerPane } from './DevelopmentServerPane';

describe('DevelopmentServerPane component', () => {
  let wrapper;
  let instance;
  lolex.install();

  const task = {
    name: 'start',
    status: 'idle',
  };

  const project = {
    id: 'a-project',
    type: 'create-react-app',
  };

  const mockActions = {
    launchDevServer: jest.fn(),
    abortTask: jest.fn(),
  };

  beforeEach(() => {
    wrapper = shallow(
      <DevelopmentServerPane task={task} project={project} {...mockActions} />
    );
    instance = wrapper.instance();
  });

  describe('Rendering', () => {
    it('should render', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should return message if no tasks', () => {
      wrapper = shallow(<DevelopmentServerPane />);
      expect(wrapper.text()).toMatch(/This project does not appear/);
    });
  });

  describe('Component logic', () => {
    it('should start devServer', () => {
      instance.handleToggle(true);
      expect(mockActions.launchDevServer).toHaveBeenCalledWith(
        task,
        new Date()
      );
    });

    it('should abort task', () => {
      instance.handleToggle(false);
      expect(mockActions.abortTask).toHaveBeenCalledWith(
        task,
        project.type,
        new Date()
      );
    });
  });
});
