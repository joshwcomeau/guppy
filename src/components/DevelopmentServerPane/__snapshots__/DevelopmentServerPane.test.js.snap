// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevelopmentServerPane component Rendering should render 1`] = `
<Pane
  moreInfoHref="https://github.com/joshwcomeau/guppy/blob/master/docs/getting-started.md#development-server"
  primaryActionChildren={
    <Toggle
      isToggled={false}
      onToggle={[Function]}
      padding={2}
      size={32}
    />
  }
  title="Development Server"
>
  <OnlyOn
    display="inline"
    size="mdMin"
  >
    <styled.div>
      <styled.div>
        <styled.div>
          Runs a local development server that updates whenever you make changes to the files.
        </styled.div>
        <DevelopmentServerStatus
          status="idle"
        />
        <styled.div>
          <ExternalLink
            color="#304FFE"
            display="inline-block"
            hoverColor="#3f6cff"
            href="https://github.com/facebook/create-react-app#user-guide"
          >
            View Documentation
          </ExternalLink>
        </styled.div>
      </styled.div>
      <styled.div>
        <Connect(TerminalOutput)
          height={300}
          task={
            Object {
              "name": "start",
              "status": "idle",
            }
          }
          title="Server Logs"
        />
      </styled.div>
    </styled.div>
  </OnlyOn>
  <OnlyOn
    display="inline"
    size="sm"
  >
    <styled.div>
      <styled.div>
        <styled.div>
          <styled.div>
            Runs a local development server that updates whenever you make changes to the files.
          </styled.div>
          <styled.div
            size={10}
          />
          <styled.div>
            <ExternalLink
              color="#304FFE"
              display="inline-block"
              hoverColor="#3f6cff"
              href="https://github.com/facebook/create-react-app#user-guide"
            >
              View Documentation
            </ExternalLink>
          </styled.div>
        </styled.div>
      </styled.div>
      <styled.div>
        <Connect(TerminalOutput)
          height={300}
          task={
            Object {
              "name": "start",
              "status": "idle",
            }
          }
        />
      </styled.div>
    </styled.div>
  </OnlyOn>
</Pane>
`;
